lib/fhir_models/validation/structure_validator.rb

```diff
@@ -0,0 +1,135 @@
+module FHIR
+  module Validation
+    # Validator which allows validation of a resource against a profile
+    class StructureValidator
+      @vs_validators = {}
+
+      attr_accessor :all_results
+      attr_accessor :show_skipped
+      attr_reader :element_validators
+
+      # Create a StructureValidator from a FHIR::StructureDefinition
+      #
+      # @param profile [FHIR::StructureDefinition]
+      #
+      def initialize(profile, use_default_element_validators: true)
+        @profile = profile
+        @show_skipped = true
+        @element_validators = Set.new
+        add_default_element_validators if use_default_element_validators
+      end
+
+      # Validate the provided resource
+      #
+      # @param resource [FHIR::Model] The Resource to be validated
+      # @return [Hash] the validation results
+      def validate(resource)
+        return [FHIR::ValidationResult.new(validation_type: :structure, result: :fail)] unless resource.is_a? FHIR::Model
+
+        validate_against_snapshot(resource)
+
+        FHIR::Validation::StructureValidationResult.new(@snapshot_hierarchy).all_results
+      end
+
+      # Removes the currently registered element validators
+      def clear_element_validators
+        @element_validators = Set.new
+      end
+
+      # Add additional element validators
+      # @param element_validator [Enumerable<#validate>]
+      def register_element_validators(element_validator)
+        @element_validators.add(*element_validator)
+      end
+
+      # Add the default Element Validators to the StructureValidator
+      def add_default_element_validators
+        @element_validators.merge(Set[Validation::CardinalityValidator,
+                                      Validation::DataTypeValidator,
+                                      Validation::FixedValueValidator,
+                                      Validation::MaxLengthValidator])
+      end
+
+      # Build a hierarchy from a list of ElementDefinitions
+      #
+      # @param [Array] List of Element Definitions
+      # @return [Hash] The ElementDefinition hierarchy
+      private def build_hierarchy(elem_def_list)
+        hierarchy = {}
+        elem_def_list.each do |element|
+          # Separate path and slices into an array of keys
+          slices = element.id.split(':')
+          path = slices.shift.split('.')
+          slices = slices.pop&.split('/')
+          last_path = slices.nil? ? path.pop : slices.pop
+
+          # Build the hierarchy
+          thing = path.inject(hierarchy) do |memo, k|
```

Comment:
If it needs a comment, it should probably be its own method. Also, `thing`, `memo`, and `k` are pretty bad names.